02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+  - if is the `opener`:
+    - MUST fail the channel if:
+      - the total count of `input_info`s is greater than 6
+
+
+##### Rationale
+Each node must have a complete set of the transaction inputs
+to derive the funding transaction and subsequent commitment signatures.
+
+`sats` is the satoshi value of the input.
+
+`prevtx_txid`, `prevtx_vout`, and `prevtx_scriptpubkey` specify the output
+to be spent.  `prevtx_txid` is the hash of the transaction that this input is
+an output of; `prevtx_vout` its outpoint; and `prevtx_scriptpubkey` the
+corresponding `scriptPubKey` value.
+
+`max_witness_len` is the total serialized length of the
+witness data that will be supplied (e.g. sizeof(varint) +
+sizeof(witness) for each) in `funding_signed2`.
+
+`input_info`.`script` is the scriptSig field for the input. Only applicable
+for P2SH-wrapped inputs.
+Native SegWit inputs (P2WPKH and P2WSH) inputs, will have an empty `script` field
+See [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#examples).
+
+The `accepter` node may omit this message.
+
+
+#### The `funding_add_output` Mesage
+
+This message contains information for outputs to a funding transaction.
+
+1. type: 60 (`funding_add_output`)
+2. data:
+    * [`32*byte`:`temporary_channel_id`]
+    * [`u16`:`num_outputs`]
+    * [`num_outputs*output_info`:`output_info`]
+
+1. subtype: `output_info`
+2. data:
+    * [`u64`:`sats`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+##### Requirements
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST NOT send `funding_add_output` if it has already transmitted `funding_add_complete`
+  - MUST send this after `accept_channel2` has been exchanged, but before `funding_add_complete`
+  - MUST add all sent outputs to the funding transaction
+  - MUST ensure the `output_info`.`script` is a standard script
+  - MUST NOT include the channel funding output.
+  - if is the `opener`:
+    - MAY specify an output with value zero, which will be used
+      as the change address for any resulting funds after fees are deducted
+  - if is the `accepter`:
+    - MUST NOT send a total count of more than 8 outputs, across all `funding_add_output` messages.
+
+The receiving node:
+  - MUST add all received outputs to the funding transaction
+  - MUST fail the channel if
+    - it receives this message after receiving `funding_add_complete`
+  - if is the `opener`:
+    - MUST fail the channel if:
+      - receives a total count of more than 3 outputs, across all `funding_add_output`s
+
+
+##### Rationale
+Each node must have a complete set of the transaction outputs
+to derive the funding transaction and subsequent commitment signatures.
+
+`sats` is the satoshi value of the output.
+
+The opening node may include one output with `sats` value
+of zero. This will be used for change, or discarded if its value is
+be below `dust_limit_satoshis`.
+
+Change is calculated as the sum of the opener's `input_info`.`sats`
```

Comment:
I'm not sure about the opener paying only the fees, it means if Alice funds with 1 input as opener but Bob puts max input 16, Bob utxo aggregation cost is externalized to Alice ? Sure you can hardcode or use a smart-fee-algo(tm) to reject but incentive-level it means accepter should seek this limit to free-ride ?

Reply:
Yess, I've got a draft for how to re-distribute fees (and a tweak to this protocol) I'm about to send to the mailing list. :)